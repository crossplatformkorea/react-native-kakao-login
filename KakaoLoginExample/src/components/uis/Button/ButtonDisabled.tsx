import type {StyleProp, TextStyle, ViewStyle} from 'react-native';

import {Body3} from '../Typography';
import type {FC} from 'react';
import React from 'react';
import {StyledButton} from '../Styles';
import styled from 'styled-components/native';

const StyledButtonDisabled = styled(StyledButton)`
  background-color: ${({theme}) => theme.disabled};
  border-color: rgb(200, 200, 200);
`;

const StyledTextDisabled = styled(Body3)`
  color: ${({theme}) => theme.disabled};
`;

type Props = {
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  text?: string;
};

const ButtonDisabled: FC<Props> = ({style, textStyle, text}) => {
  return (
    <StyledButtonDisabled style={style}>
      <StyledTextDisabled style={textStyle}>{text}</StyledTextDisabled>
    </StyledButtonDisabled>
  );
};

export default ButtonDisabled;
