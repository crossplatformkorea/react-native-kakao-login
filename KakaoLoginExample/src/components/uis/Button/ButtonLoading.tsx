import type {ActivityIndicatorProps, StyleProp, ViewStyle} from 'react-native';

import {ActivityIndicator} from 'react-native';
import type {FC} from 'react';
import React from 'react';
import {StyledButton} from '../Styles';

type Props = {
  style?: StyleProp<ViewStyle>;
  indicatorColor?: ActivityIndicatorProps['color'];
  indicatorSize?: ActivityIndicatorProps['size'];
};

const ButtonLoading: FC<Props> = ({
  style,
  indicatorColor,
  indicatorSize = 'small',
}) => {
  return (
    <StyledButton style={style}>
      <ActivityIndicator size={indicatorSize} color={indicatorColor} />
    </StyledButton>
  );
};

export default ButtonLoading;
