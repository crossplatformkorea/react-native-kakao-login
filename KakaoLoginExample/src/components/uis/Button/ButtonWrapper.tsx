import type {
  ImageSourcePropType,
  ImageStyle,
  StyleProp,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

import {Body3} from '../Typography';
import type {FC} from 'react';
import React from 'react';
import {StyledButton} from '../Styles';
import {TouchableOpacity} from 'react-native';
import styled from 'styled-components/native';

const StyledImage = styled.Image`
  width: 24px;
  height: 24px;
  position: absolute;
  left: 16px;
`;

const StyledText = styled(Body3)`
  color: ${({theme}) => theme.textContrast};
`;

type Props = {
  testID?: TouchableOpacityProps['testID'];
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  onPress?: TouchableOpacityProps['onPress'];
  imgLeftSrc?: ImageSourcePropType;
  imgLeftStyle?: StyleProp<ImageStyle>;
  style?: ViewStyle;
  textStyle?: TextStyle;
  text?: string;
};

const Button: FC<Props> = ({
  testID,
  activeOpacity = 0.7,
  onPress,
  imgLeftSrc,
  imgLeftStyle,
  style,
  textStyle,
  text = '',
}) => {
  return (
    <TouchableOpacity
      testID={testID}
      activeOpacity={activeOpacity}
      onPress={onPress}>
      <StyledButton style={style}>
        {imgLeftSrc ? (
          <StyledImage style={imgLeftStyle} source={imgLeftSrc} />
        ) : null}
        <StyledText style={textStyle}>{text}</StyledText>
      </StyledButton>
    </TouchableOpacity>
  );
};

export default Button;
