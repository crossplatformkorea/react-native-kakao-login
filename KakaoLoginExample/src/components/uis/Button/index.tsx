import type {
  ImageSourcePropType,
  ImageStyle,
  StyleProp,
  TextStyle,
  ViewStyle,
} from 'react-native';

import ButtonDisabled from './ButtonDisabled';
import ButtonLoading from './ButtonLoading';
import ButtonWrapper from './ButtonWrapper';
import type {FC} from 'react';
import React from 'react';

interface Props {
  testID?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  onPress?: () => void;
  style?: ViewStyle;
  disabledStyle?: ViewStyle;
  textStyle?: TextStyle;
  imgLeftSrc?: ImageSourcePropType;
  imgLeftStyle?: StyleProp<ImageStyle>;
  indicatorColor?: string;
  activeOpacity?: number;
  text?: string;
}

const Button: FC<Props> = ({
  testID,
  isLoading,
  isDisabled,
  onPress,
  style,
  disabledStyle,
  textStyle,
  imgLeftSrc,
  imgLeftStyle,
  indicatorColor,
  activeOpacity,
  text,
}) => {
  if (isDisabled)
    return <ButtonDisabled style={disabledStyle} textStyle={textStyle} />;

  if (isLoading) return <ButtonLoading indicatorColor={indicatorColor} />;

  return (
    <ButtonWrapper
      testID={testID}
      onPress={onPress}
      imgLeftSrc={imgLeftSrc}
      style={style}
      imgLeftStyle={imgLeftStyle}
      activeOpacity={activeOpacity}
      text={text}
      textStyle={textStyle}
    />
  );
};

export default Button;
