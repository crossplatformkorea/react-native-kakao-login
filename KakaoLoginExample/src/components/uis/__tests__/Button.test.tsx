import 'react-native';

import * as React from 'react';

import {
  RenderAPI,
  act,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react-native';

import Button from '../Button';
import {ThemeType} from 'dooboo-ui';
import {createTestElement} from '../../../../test/testUtils';

// Note: test renderer must be required after react-native.

let props: any;
let component: React.ReactElement;
let testingLib: RenderAPI;

describe('[Button]', () => {
  let cnt = 1;

  beforeEach(() => {
    props = {
      onPress: (): number => cnt++,
      testID: 'btn',
    };

    component = createTestElement(<Button {...props} />);
  });

  afterEach(cleanup);

  it('[ThemeType.Light] renders without crashing', () => {
    testingLib = render(component);

    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  it('should render [ThemeType.Dark] without crashing', () => {
    component = createTestElement(<Button {...props} />, ThemeType.DARK);

    testingLib = render(component);

    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  it('should render [isDisabled] status without crashing', () => {
    props.isDisabled = true;
    component = createTestElement(<Button {...props} />);

    testingLib = render(component);

    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  it('should render [isLoading] status without crashing', () => {
    props.isLoading = true;
    component = createTestElement(<Button {...props} />);

    testingLib = render(component);

    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  describe('interactions', () => {
    beforeEach(() => {
      testingLib = render(component);
    });

    afterEach(cleanup);

    it('should simulate onClick', () => {
      const btn = testingLib.queryByTestId('btn');

      act(() => {
        fireEvent.press(btn);
        fireEvent.press(btn);
      });

      expect(cnt).toBe(3);
    });
  });
});
