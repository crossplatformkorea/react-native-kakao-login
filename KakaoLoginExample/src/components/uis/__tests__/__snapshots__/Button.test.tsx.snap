// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] [ThemeType.Light] renders without crashing 1`] = `
<View
  activeOpacity={0.7}
  onPress={[Function]}
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#393D7A",
          "borderColor": "#393D7A",
          "borderRadius": 4,
          "borderWidth": 2,
          "height": 52,
          "justifyContent": "center",
          "width": 320,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#D3D8E8",
            "fontSize": 14,
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`[Button] should render [ThemeType.Dark] without crashing 1`] = `
<View
  activeOpacity={0.7}
  onPress={[Function]}
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#8A96DC",
          "borderColor": "#8A96DC",
          "borderRadius": 4,
          "borderWidth": 2,
          "height": 52,
          "justifyContent": "center",
          "width": 320,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 14,
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`[Button] should render [isDisabled] status without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#00000030",
        "borderColor": "rgb(200, 200, 200)",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#00000030",
          "fontSize": 14,
        },
      ]
    }
  />
</View>
`;

exports[`[Button] should render [isLoading] status without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#393D7A",
        "borderColor": "#393D7A",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;
